src/contradictory_claims/models/sbert_models.py

```diff
@@ -0,0 +1,382 @@
+    :type df_val: pd.DataFrame()
+    :param epochs: numer of epochs
+    :type epochs: int
+    :param learning_rate: learning rate
+    :type learning_rate: float
+    :param batch_size: batch size to be used for training
+    :type batch_size: int
+    """
+    if embedding_epochs is None:
+        embedding_epochs=epochs
+    nli_reader = NLIDataReader(df_train)
+    train_num_labels = nli_reader.get_num_labels()
+
+    train_data = SentencesDataset(nli_reader.get_examples(), model=model.embedding_model)
+    train_data.label_type = torch.long
+    # some bug in sentence_transformer library causes it to be identified as float by default
+    train_dataloader_embed = DataLoader(train_data, shuffle=True, batch_size=batch_size)
+    train_loss_embed = losses.SoftmaxLoss(
+        model=model.embedding_model,
+        sentence_embedding_dimension=model.embedding_model.get_sentence_embedding_dimension(),
+        num_labels=train_num_labels)
+
+    val_nli_reader = NLIDataReader(df_val)
+    dev_data = SentencesDataset(val_nli_reader.get_examples(), model=model.embedding_model)
+    dev_data.label_type = torch.long
+    evaluator = EmbeddingSimilarityEvaluator(sentences1=df_val["sentence1"].values,
+                                             sentences2=df_val["sentence2"].values,
+                                             scores=df_val["label"].values / 2.,
+                                             batch_size=batch_size)
+    warmup_steps = math.ceil(len(train_dataloader_embed) * epochs / batch_size * 0.1)  # 10% of train data for warm-up
+
+    # now to train the final layer
+    train_dataset = ClassifierDataset(df_train, tokenizer=tokenizer)
+    val_dataset = ClassifierDataset(df_val, tokenizer=tokenizer)
+    if enable_class_weights==False:
+        class_weights=None
+    else:
+        class_weights = train_dataset.class_weights()
+
+    train_dataloader = DataLoader(dataset=train_dataset,
+                                  batch_size=batch_size,
+                                  collate_fn=collate_fn,
+                                  shuffle=True)
+    val_dataloader = DataLoader(dataset=val_dataset,
+                                batch_size=1,
+                                collate_fn=collate_fn)
+
+    device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+    criterion = nn.CrossEntropyLoss(weight=class_weights.to(device))
+    optimizer = optim.Adam(model.parameters(), lr=learning_rate)
+    model.to(device)
+    accuracy_stats = {"train": [],
+                      "val": [],
+                      }
+    loss_stats = {"train": [],
+                  "val": [],
+                  }
+
+    print("------TRAINING STARTS----------")  # noqa: T001
+    ## train embedding layer
+    unfreeze_layer(model.embedding_model)
+    model.embedding_model.fit(train_objectives=[(train_dataloader_embed, train_loss_embed)],
+                              evaluator=evaluator,
+                              epochs=1,
```

Comment:
Should it be "'epochs=epochs"?

Reply:
according to SBERT paper they achieved best results when using epochs=1 and lr=2e-5, I was still playing around with what best to keep, should it be different epochs for SBERT architecture and classification layer or the same